### PR TITLE
Adding allowed_resource_types policy

### PR DESCRIPTION
**Pull request by AdrienWalkowiak**
This template goal is to allow a user to specify which resources types are allowed (whitelist) or denied (blacklist) in their GCP infrastructure (for instance within a folder). 

This can be used to restrict what kind of resources should live in a folder, or be deployed by a pipeline (for instance having separate pipelines for your network vs compute resources etc.).

### CODE REVIEW

File: validator/allowed_resource_types_test.rego [hunk 1]
```diff
@@ -0,0 +1,92 @@
+#
+# Copyright 2018 Google LLC
```
Comment (jdyke) [inline validator/allowed_resource_types_test.rego]: can we update this to 2019?

Comment (AdrienWalkowiak) [reply validator/allowed_resource_types_test.rego]: done - I also fixed the sample file name

File: policies/templates/gcp_dataproc_location_v1.yaml [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# Copyright 2019 Google LLC
```
Comment (joecheuk) [inline policies/templates/gcp_dataproc_location_v1.yaml]: Is this change intended to be in this PR?

File: policies/templates/gcp_allowed_resource_types_v1.yaml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            resource_types_list:
+              type: array
+              items: string
+              description: "Array of resource types that will be either
+              authorized (mode: whitelist) or denied (mode: blacklist)."
+
+  targets:
+    validation.gcp.forsetisecurity.org:
+      rego: | #INLINE("validator/allowed_resource_types.rego")
+       #
```
Comment (joecheuk) [inline policies/templates/gcp_allowed_resource_types_v1.yaml]: The indentation doesn't look quite right, does this get parsed properly in config validator?

File: policies/templates/gcp_dataproc_location_v1.yaml (hunk 2 above)
Comment (AdrienWalkowiak) [reply policies/templates/gcp_dataproc_location_v1.yaml]: it was not originally, but merging both policies will fix my branches on my fork for future PRs. Sorry for the confusion.

Comment (AdrienWalkowiak) [reply policies/templates/gcp_dataproc_location_v1.yaml]: I removed it from this PR

File: policies/templates/gcp_allowed_resource_types_v1.yaml (hunk 3 above)
Comment (AdrienWalkowiak) [reply policies/templates/gcp_allowed_resource_types_v1.yaml]: Fixed it - yamlint was happy with it, but better stay consistent

File: validator/allowed_resource_types.rego [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	asset := input.asset
+
+	# Retrieve the current mode if passed, use "whitelist" as a default
+	mode := lower(lib.get_default(params, "mode", "whitelist"))
+
+	# Retrieve the resource types list - default to empty set
+	ressource_types_list := cast_set(lib.get_default(params, "resource_types_list", {}))
+
+	# The asset raises a violation if resource_type_is_valid is evaluated to false (both of them)
+not resource_type_is_valid(	mode, asset, ressource_types_list)
```
Comment (joecheuk) [inline validator/allowed_resource_types.rego]: This line has some formatting issue.

File: validator/allowed_resource_types.rego [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+resource_type_is_valid(mode, asset, resource_types_list) {
+	# "if we are in a blacklist mode"
+	mode == "blacklist"
+
+	# Retrieve the asset type
+	asset_type := asset.asset_type
+
+	#the asset is valid only if it's not in the resource_type_list
```
Comment (joecheuk) [inline validator/allowed_resource_types.rego]: nit: `# the`

File: validator/allowed_resource_types.rego [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+###########################
+
+resource_type_is_valid(mode, asset, resource_types_list) {
+	# anything other than blacklist is treated as "whitelist"
+	mode != "blacklist"
+
+	# Retrieve the asset type
+	asset_type := asset.asset_type
+
+	#the asset is valid if it's in the resource_type_list
```
Comment (joecheuk) [inline validator/allowed_resource_types.rego]: nit: `# the`

File: validator/allowed_resource_types.rego [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+resource_type_is_valid(mode, asset, resource_types_list) {
+	# "if we are in a blacklist mode"
+	mode == "blacklist"
+
+	# Retrieve the asset type
+	asset_type := asset.asset_type
+
+	#the asset is valid only if it's not in the resource_type_list
+not 	resource_types_list[asset_type]
```
Comment (joecheuk) [inline validator/allowed_resource_types.rego]: This line has some formatting issue.

File: policies/templates/gcp_allowed_resource_types_v1.yaml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+          properties:
+            mode:
+              type: string
+              enum: [whitelist, blacklist]
+              description: "String identifying the operational mode,
+              whitelist or blacklist. In the whitelist mode, only resource types
+              from the resource_types_list will be allowed (all other types
+              will raise a violation). In the blacklist mode, any resource
+              type not in the resource_types_list will not raise a violation."
+            resource_types_list:
```
Comment (joecheuk) [inline policies/templates/gcp_allowed_resource_types_v1.yaml]: maybe `resource_type_list` is better?

File: validator/allowed_resource_types.rego (hunk 5 above)
Comment (AdrienWalkowiak) [reply validator/allowed_resource_types.rego]: updated

File: validator/allowed_resource_types.rego (hunk 6 above)
Comment (AdrienWalkowiak) [reply validator/allowed_resource_types.rego]: updated

File: policies/templates/gcp_allowed_resource_types_v1.yaml (hunk 8 above)
Comment (AdrienWalkowiak) [reply policies/templates/gcp_allowed_resource_types_v1.yaml]: renamed to resource_type_list. Thanks, I was not sure about it
